File: day30/alienDictionary.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <string>

using namespace std;

string alienOrder(const vector<string>& words) {
    unordered_map<char, unordered_set<char>> graph;
    unordered_map<char, int> inDegree;

    // Initialize in-degree for all unique characters
    for (const string& word : words) {
        for (char c : word) {
            inDegree[c] = 0;
        }
    }

    // Build graph edges and update in-degrees
    for (size_t i = 0; i < words.size() - 1; ++i) {
        string first = words[i];
        string second = words[i + 1];
        int len = min(first.size(), second.size());

        if (first.size() > second.size() && first.substr(0, len) == second) {
            return "";
        }

        for (int j = 0; j < len; ++j) {
            char a = first[j];
            char b = second[j];
            if (a != b) {
                if (!graph[a].count(b)) {
                    graph[a].insert(b);
                    inDegree[b]++;
                }
                break;
            }
        }
    }

    // Topological sort (BFS)
    queue<char> q;
    for (const auto& [c, degree] : inDegree) {
        if (degree == 0) q.push(c);
    }

    string order;
    while (!q.empty()) {
        char c = q.front(); q.pop();
        order += c;
        for (char neighbor : graph[c]) {
            if (--inDegree[neighbor] == 0) q.push(neighbor);
        }
    }

    if (order.length() != inDegree.size()) return "";

    return order;
}

int main() {
    vector<string> words = {"wrt", "wrf", "er", "ett", "rftt"};
    
    cout << "Input words:\n";
    for (const string& word : words) {
        cout << word << "\n";
    }

    string order = alienOrder(words);

    cout << "\nInferred alien dictionary order:\n";
    if (order.empty()) {
        cout << "(Invalid dictionary — cycle or prefix conflict detected)\n";
    } else {
        cout << order << "\n";
    }

    return 0;
}

File: day30/AlienDictionary.java
import java.util.*; 

public class AlienDictionary {

    public static String alienOrder(String[] words) {
        // Step 1: Create data structures
        Map<Character, Set<Character>> graph = new HashMap<>();
        Map<Character, Integer> inDegree = new HashMap<>();

        // Initialize inDegree for all unique characters
        for (String word : words) {
            for (char c : word.toCharArray()) {
                inDegree.putIfAbsent(c, 0);
            }
        }

        System.out.println("Initial in-degree map: " + inDegree);

        // Step 2: Build graph by comparing adjacent words
        for (int i = 0; i < words.length - 1; i++) {
            String first = words[i];
            String second = words[i + 1];
            int len = Math.min(first.length(), second.length());

            if (first.length() > second.length() && first.startsWith(second)) {
                // Invalid case: prefix problem
                return "";
            }

            for (int j = 0; j < len; j++) {
                char a = first.charAt(j);
                char b = second.charAt(j);
                if (a != b) {
                    graph.putIfAbsent(a, new HashSet<>());
                    if (!graph.get(a).contains(b)) {
                        graph.get(a).add(b);
                        inDegree.put(b, inDegree.get(b) + 1);
                    }
                    break;
                }
            }
        }

        // Debug print of graph
        System.out.println("Graph (edges):");
        for (char key : graph.keySet()) {
            System.out.println("  " + key + " → " + graph.get(key));
        }
        System.out.println("Updated in-degree map: " + inDegree);

        // Step 3: Topological sort using BFS
        Queue<Character> queue = new LinkedList<>();
        for (char c : inDegree.keySet()) {
            if (inDegree.get(c) == 0) {
                queue.offer(c);
            }
        }

        StringBuilder result = new StringBuilder();
        System.out.println("Starting BFS with queue: " + queue);

        while (!queue.isEmpty()) {
            char current = queue.poll();
            result.append(current);
            System.out.println("Visiting: " + current);

            if (graph.containsKey(current)) {
                for (char neighbor : graph.get(current)) {
                    inDegree.put(neighbor, inDegree.get(neighbor) - 1);
                    if (inDegree.get(neighbor) == 0) {
                        queue.offer(neighbor);
                    }
                }
            }
        }

        // If the result does not include all characters, there was a cycle
        if (result.length() < inDegree.size()) {
            return "";
        }

        return result.toString();
    }

    public static void main(String[] args) {
        String[] words = {"wrt", "wrf", "er", "ett", "rftt"};
        System.out.println("Input words: " + Arrays.toString(words));
        System.out.println("Alien dictionary order: " + alienOrder(words));
    }
}
